### PR TITLE
Relay maintainer: Push blocks happy path

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Depends on: #765

Here we introduce the block forwarder pushing loop which supports a basic case of pushing a batch of headers within which the chain difficulty doesn't change. Specifically, this pull request:
- Adds a general structure of the block forwarder
- Introduces forwarder lifecycle logic and error handling
- Implements pulling block headers from the internal headers queue 
- Adds pushing block headers to the host chain in case there is no difficulty change 
- Implements updating best header if needed
- Performs some steps to improve the work with raw bytes by introducing `btc.Digest` type
- Improves documentation of chain interfaces

### CODE REVIEW

File: relay/pkg/chain/chain.go
Comment (pdyraga) [inline relay/pkg/chain/chain.go:13]: Should we declare `type Digest [32]byte` just like we do for BTC? There is a lot of `[32]byte` declarations in this file.

File: relay/pkg/node/node.go
Comment (pdyraga) [inline relay/pkg/node/node.go:17]: Agreed. 😆  Though this is already happening, it's not a TODO?

File: relay/pkg/node/node.go
Comment (pdyraga) [inline relay/pkg/node/node.go:50]: Maybe just `"relay occurred an error: [%v]"`? It will be clear from the logs it is continuing its work. "Restart" feels a little too rough. As an operator, I'd expect the entire process is restarted/

File: relay/pkg/btc/remote/remote.go
Comment (pdyraga) [inline relay/pkg/btc/remote/remote.go:97]: I have already noted it and I know I am annoying but `btc/remote/remote.go` and `btc/local/local.go` feels wrong to me especially that they are exported. We can, for example, do `btc/client.go` and `btc/local_client.go` and have two functions, `Connect()` and `ConnectLocal()`

Comment (pdyraga) [reply relay/pkg/btc/remote/remote.go:97]: Not in this PR of course, just grumbling. 😁

File: relay/pkg/node/node.go
Comment (pdyraga) [inline relay/pkg/node/node.go:52]: But what does it mean to the operator? Maybe just `"stopping the relay"`?

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [inline relay/pkg/block/forwarder.go:72]: I don't think we need it, especially that it is already logged one level higher. We can log it here or in the code calling `RunForwarder` but let's do it only one time.

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [inline relay/pkg/block/forwarder.go:89]: Instead of pushing the error to the channel and hoping that the control loop will restart this process, why don't we just log an error and continue?

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [inline relay/pkg/block/forwarder.go:95]: I'd rather have `pullHeadersFromQueue` blocking until it has as least one header to return. Feels more elegant.

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [inline relay/pkg/block/forwarder.go:36]: Thoughts on renaming it to `Relay`?

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [inline relay/pkg/block/forwarder.go:89]: Let's stay that `pushHeadersToHostChain` failed. If we `return` and restart this loop or just log an error and `continue` without doing anything else, `headers` will not be pushed and we'll be waiting for another batch of headers from `pullHeadersFromQueue`. Is it blocking the maintainer on-chain? Can we afford to drop one batch of headers?

File: relay/pkg/block/push.go
Comment (pdyraga) [inline relay/pkg/block/push.go:14]: `... and returns them from the function`?

File: relay/pkg/block/push.go
Comment (pdyraga) [inline relay/pkg/block/push.go:15]: Thoughts on naming it `pullHeaders` or `pullBtcHeaders`? The fact we take them from `f.headersQueue` is an implementation details. From the perspective of the code calling this function, we are pulling them from Bitcoin chain.

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [reply relay/pkg/block/forwarder.go:95]: Wait, this is already happening. Why do we need to suspend the forwarder? Wouldn't it be enough to increase the timeout?

File: relay/pkg/block/push.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/push.go:15]: I'm rather against it because within the forwarder we pull block either from the BTC chain or from the queue and naming it just `pullHeaders` will be ambiguous. I think it's more readable and consistent if we have on the push side:
- `pullHeadersFromQueue`
- `pushHeadersToHostChain`
and on the pull side (see https://github.com/keep-network/tbtc/pull/769):
- `pullHeadersFromBtcChain`
- `pushHeadersToQueue`

Last but not least, those methods are package-private helpers, the client code just do `block.RunForwarder` and doesn't care.

File: relay/pkg/chain/chain.go
Comment (lukasz-zimnoch) [reply relay/pkg/chain/chain.go:13]: I thought about this but this will require conversions between `btc.Digest` and `chain.Digest` in the `block.Forwarder`. That seemed odd to me as `block.Forwarder` shouldn't care about data conversions but just forward headers according to defined logic. Of course, this is a simple conversion but still, I'm not 100% sure.

File: relay/pkg/node/node.go
Comment (lukasz-zimnoch) [reply relay/pkg/node/node.go:17]: It's not done as we don't support new features yet. What we have done is just forwarder which can be disabled using contexts and is restarted upon failures.

File: relay/pkg/btc/remote/remote.go
Comment (lukasz-zimnoch) [reply relay/pkg/btc/remote/remote.go:96]: Yeah, that's probably a good move. I'll leave a TODO and address this in a follow-up PR.

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:89]: Because we can enter into an inconsistent state. It's far better to fail here, restart the forwarder, initialize it with some chain data (it's needed by the pull side), and start from the beginning. This is also the way it was handled in the legacy relay maintainer but there the entire app crashed. Here we just restart the forwarding part.

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:95]: I think there is no sense in continuously looping and checking timeouts on the queue in case there is no new headers. Also, if there are more blocks on the queue it's more efficient to handle them in batch than pulling them one by one. This is also how the legacy maintainer works.

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:36]: I'm not sure as the entire subdirectory is named `Relay`, we have a `Relay` contract and we have `chain.Relay` 😁  I think `Forwader` is a better name for things we are doing here. I always considered `Relay` to be a high-level concept which consists of BTC and ETH connectivity, block forwarder, and auxiliary stuff.

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:89]: It's even better to `return`, restart, and drop the batch entirely. In that case, the new forwarder instance will initialize the pull side properly basing on `Relay` contract state. If we just log and continue, we will receive the next batch from the queue which cannot be pushed on the host chain as the chain still "waits" for the old batch. The forwarder will remain in an inconsistent state.

File: relay/pkg/node/node.go
Comment (lukasz-zimnoch) [reply relay/pkg/node/node.go:47]: Agree however because of https://github.com/keep-network/tbtc/pull/767#discussion_r598995537 my proposition is: https://github.com/keep-network/tbtc/commit/c6fb5d71e9c976fb98ee676e36dfcd48ff6221fe

File: relay/pkg/node/node.go
Comment (lukasz-zimnoch) [reply relay/pkg/node/node.go]: Agree however because of https://github.com/keep-network/tbtc/pull/767#discussion_r598995537 my proposition is: https://github.com/keep-network/tbtc/commit/c6fb5d71e9c976fb98ee676e36dfcd48ff6221fe

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go]: Done in https://github.com/keep-network/tbtc/commit/c6fb5d71e9c976fb98ee676e36dfcd48ff6221fe

File: relay/pkg/block/push.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/push.go]: Please see https://github.com/keep-network/tbtc/commit/c6fb5d71e9c976fb98ee676e36dfcd48ff6221fe

**Comment by lukasz-zimnoch** [comment]
@pdyraga ready for another look!

File: relay/pkg/btc/remote/remote.go
Comment (lukasz-zimnoch) [reply relay/pkg/btc/remote/remote.go:96]: Follow-up PR: https://github.com/keep-network/tbtc/pull/770

File: relay/pkg/chain/chain.go
Comment (pdyraga) [reply relay/pkg/chain/chain.go:13]: I think this conversion feels wrong because we would be converting from BTC header to... BTC header but declared in the host chain package. Instead of having two (identical) representations of BTC header, maybe the interface from `pkg/chain/chain.go` should import types from `pkg/btc/btc.go`?

**Review by tomaszslabon — APPROVED** [review]

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:87]: Please see https://github.com/keep-network/tbtc/commit/425f216ae8171d841a9791b6e752581258481412

File: relay/pkg/node/node.go
Comment (pdyraga) [reply relay/pkg/node/node.go:17]: Ah, I see - we are talking about relay auctions. Yes, it makes sense. Let's make it clear in the TODO what we are talking about. Proposition:
```
// TODO: This function will be probably the right place to handle relay auctions
// which will require starting and stopping the block forwarder.
```

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [reply relay/pkg/block/forwarder.go:36]: I think this is fine, it is the same situation as with `keep-core/pkg/beacon` or `keep-ecdsa/pkg/ecdsa` where the "heart" package of the client is named after the protocol. There is much more logic happening inside this package than just forwarding: `updateBestHeader`, `findLastCommonAncestor` - this code is about how the relay works.

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [inline relay/pkg/block/forwarder.go:32]: I'd go with `tbtc-relay` given that the repository is `tbtc`. We have `keep-beacon` logger for `keep-core/pkg/beacon` and `keep-tss` logger for `keep-ecdsa/pkg/ecdsa/tss`.

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [inline relay/pkg/block/forwarder.go:51]: If we decide to rename the package to `relay`, I propose we name it `Start` and make it clear in the docs that the relay exits on a first error. This will explain the logic in `runForwarderControlLoop` in `node.go`.

**Review by pdyraga — COMMENTED** [review]
The first portion, more to come depending on their outcome.

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:32]: I think it will be too long. We will have `tbtc-relay-block-forwarder`. The `tbtc` prefix doesn't give any value in my opinion. No need to know we are in the `tbtc` repo from the `relay`'s perspective.

Comment (pdyraga) [reply relay/pkg/block/forwarder.go:32]: Per another comment, I'd really love to rename it to "relay" and then it's just `tbtc-relay`.

Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:32]: So, maybe just `relay` then. Still not getting the point of forcing `tbtc` prefix here 😁

Comment (pdyraga) [reply relay/pkg/block/forwarder.go:32]: To keep it consistent with keep client logging where all logs are prefixed with `keep-*` and it may help with filtering logs.
I was even thinking about `keep-relay` but the relay is a part of tBTC, not Keep Network.

Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:32]: `tbtc` repo is part of `keep-network` organization so `keep-relay` is not that bad. But ok, `tbtc-relay` may also be the option.

File: relay/pkg/chain/chain.go
Comment (lukasz-zimnoch) [reply relay/pkg/chain/chain.go:15]: Yeah, importing types from `btc.go` makes sense. However I will do this in a follow-up PR once the whole stack is merged. Left a TODO for now: https://github.com/keep-network/tbtc/commit/624b6750d90127716326f460c7c934a16f9125de

File: relay/pkg/node/node.go
Comment (lukasz-zimnoch) [reply relay/pkg/node/node.go]: See https://github.com/keep-network/tbtc/commit/551cf7c8e780374c860a4edc9c286579705867f7

File: relay/pkg/chain/chain.go
Comment (pdyraga) [reply relay/pkg/chain/chain.go:15]: Yeah, separate PR makes sense :+1:

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:36]: Technically, `updateBestHeader` and `findLastCommonAncestor` are parts of the forwarding logic as well. Also, If we rename `block.Forwarder` to `relay.Relay` then according to your comment, the future auction stuff should be also placed in `relay` package as it's how the relay will work. This seems wrong to me. There is also the thing that `relay.Relay` will use `chain.Relay` which seems weird. Finally, I think `keep-core` and `keep-ecdsa` cases are not the same. The `beacon` package is something different than the whole repo name. Same in `keep-ecdsa` where the core logic is placed in `ecdsa/tss`.

Comment (pdyraga) [reply relay/pkg/block/forwarder.go:36]: > Technically, updateBestHeader and findLastCommonAncestor are parts of the forwarding logic as well.

Right, and that's one of the things that are IMO confusing because it is not just forwarding from one place to another (just like HTTP forward for instance).

> Also, If we rename block.Forwarder to relay.Relay then according to your comment, the future auction stuff should be also placed in relay package as it's how the relay will work

No, the actions should be placed in `node.go` (or in `auctions` package referenced from `node.go`) and depending on the result of the auction, the client will start the relay or not.

> There is also the thing that relay.Relay will use chain.Relay which seems weird.

IMO it's consistent - we have a relay off-chain logic that is using relay on-chain implementation. `chain.Relay` is just an on-chain part of the interface.

> Finally, I think keep-core and keep-ecdsa cases are not the same. The beacon package is something different than the whole repo name. Same in keep-ecdsa where the core logic is placed in ecdsa/tss.

For the beacon, it's a kind of a historical decision that "core" is beacon + grant and staking contracts. If we had a separate repository for grant and staking contracts, the repository with the client would be named `keep-beacon` (this may still happen one day). For `keep-ecdsa`, we have a separate `tss` directory because we use third party implementation, but if we were using our own, it would live in `ecdsa`.

My order of preferences is:
- `pkg/relay` with `relay.go` and `Start()`
- `pkg/headers` with `headers.go` and  `StartRelay()`or `Transmit()`
- `pkg/headers` with `headers.go` and `RunForwarder()`

I don't think `pkg/block` is a good choice given that we work on headers and I don't think `forwarder.go` is a good choice because we do much more than just forwarding from one place to another.

File: relay/pkg/block/push.go
Comment (pdyraga) [inline relay/pkg/block/push.go:15]: Can we separate pull functions to `pull.go`? I kinda like that we have a separate file for push logic instead of keeping it all in one big file but having a separation of push+pull+glue code would be even more awesome.

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:36]: Yeah, I agree with the remark about `pkg/block`, this was a starting point which now can be narrowed to `pkg/header`. In that case, we can make `header.Relay` placed in `pkg/header/relay.go` exposing a `StartRelay` function. This is a small mix of your propositions but I think it's the best reflection of what we are doing here. Of course, I propose to do this as a follow-up PR after the entire stack lands on master. I'll leave a TODO. WDYT?

File: relay/pkg/block/push.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/push.go:15]: `pull.go` is about pulling blocks from BTC chain and pushing them to queue (see https://github.com/keep-network/tbtc/pull/769). `push.go` is about pulling blocks from queue and pushing to the host chain. If we move `pullHeadersFromQueue` to `pull.go` we will make a small mess here.

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [reply relay/pkg/block/forwarder.go:36]: Yeah, that's the best outcome we could get here IMO when I look at it again 👌 . Let's leave a TODO.

Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:41]: Here it is: https://github.com/keep-network/tbtc/commit/bce322f8b5881c410789c9c5f7a885815dbcc5f2

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:56]: We'll make it `StartRelay` as per https://github.com/keep-network/tbtc/pull/767#discussion_r603913978

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [reply relay/pkg/block/forwarder.go:37]: Resolving, this will be done in a separate PR according to the TODO: bce322f8b5881c410789c9c5f7a885815dbcc5f2

File: relay/pkg/node/node.go
Comment (lukasz-zimnoch) [reply relay/pkg/node/node.go:47]: As stated in https://github.com/keep-network/tbtc/pull/767#discussion_r603913978, the message will be focused around `headers` not `blocks`

File: relay/pkg/node/node.go
Comment (lukasz-zimnoch) [reply relay/pkg/node/node.go]: As stated in https://github.com/keep-network/tbtc/pull/767#discussion_r603913978, the message will be focused around `headers` not `blocks`

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [reply relay/pkg/block/forwarder.go:102]: I have totally missed that forwarder is stateful. Can we leave a comment with an explanation? My proposition:
```
				f.errChan <- fmt.Errorf("could not push headers: [%v]", err)
				// We exit on the first error letting the code controlling the
				// relay to restart it. The relay is stateful and it is easier
				// to fetch the most recent information from BTC after the
				// restart instead of trying to recover here.
				return
```

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:61]: Nope, `loopExitHandler` will be also called from the pulling loop (see #769). It's a mechanism that allows to completely shut down the entire forwarder in case one of the loops (pulling or pushing) raises an error.

File: relay/pkg/block/push.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/push.go:15]: I like this idea! Do you mind I also leave a TODO for that and address it separately as I don't have `pull.go` in this PR?

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [inline relay/pkg/block/forwarder.go:76]: Thoughts on renaming this function to `relayLoop`? (in a separate PR)

`pushingLoop` is confusing because we both pull and push inside this function.

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:106]: Sure. Please see: https://github.com/keep-network/tbtc/commit/71c059077a00c995b8eff208908ef2788873a92d

File: relay/pkg/block/push.go
Comment (pdyraga) [inline relay/pkg/block/push.go:20]: Let's clarify the case when the timeout is hit. I was confused when reviewing the code using this function thinking no headers (empty slice) can be returned in the case of a timeout. My proposition:
```
// pullHeadersFromQueue blocks until there is `headersBatchSize` headers in
// the queue or until `headerTimeout` is hit and no more headers are available
// in the queue. Function returns headers from the queue but not less than one
// and no more than `headersBatchSize` headers.
```

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [reply relay/pkg/block/forwarder.go:76]: Nevermind, this is push to host chain, other comments clarified it.

Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:76]: If we apply changes proposed in https://github.com/keep-network/tbtc/pull/767#discussion_r603954935 we will have `getHeadersFromQueue` and `pushHeadersToHostChain` so I think it won't be confusing anymore as pull/push naming will be only related with chains. Also `pushingLoop` fits better with the `pullingLoop` added in #769.

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [reply relay/pkg/block/forwarder.go:61]: In this case, what do you think about passing cancel function as a parameter? It is cleaner approach IMO: we pass the context and a function to cancel it, just like it is returned by `context.WithCancel`:

```
pushingLoop(ctx context.Context, cancelFn context.CancelFn) 
```

File: relay/pkg/block/push.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/push.go]: Sure: https://github.com/keep-network/tbtc/commit/49448a1e7de6aece3ab985d40829ddfbed1346d1 . Keep in mind this function can return an empty slice if the context is cancelled. Flagged that in the comment.

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:61]: I wanted to avoid this because I think the loop itself should not decide about cancelling its context. In my opinion, it's the responsibility of the forwarder/relay itself as it controls the lifecycle of pulling/pushing loops. What the loop should do is to inform the forwarder about an abnormal exit and the forwarder should decide what to do - in this case, cancel the `loopCtx` but in other situation, it may want to restart the loop. Also, having `loopExitHandler` we can easily add other behavior which should take place on loop exit if desired.

Comment (pdyraga) [reply relay/pkg/block/forwarder.go:61]: So maybe:
```
	go func() {
		forwarder.pushingLoop(loopCtx)
		cancelLoopCtx() // loop exited, cancel the contet
	}
```
?

I don't like we need to trust `pushingLoop` to call `loopExitHandler`. I do agree that it should not cancel the context, so the approach when the context is cancelled straight from `RunForwarder` in case one of the loops (currently we have just one loop, soon we'll have two) exits is IMO better - we do not need to trust those functions, everything is controlled by `RunForwarder` which triggered them.

Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:59]: Yep, that's a much better one. Please see: https://github.com/keep-network/tbtc/commit/37c633f2384860b280f68b77636116d214e09584

**Review by pdyraga — APPROVED** [review]

**Comment by lukasz-zimnoch** [comment]
Heads-up: A follow-up PR addressing outstanding change requests is here https://github.com/keep-network/tbtc/pull/779

**Pull request merged**
